optuna/study/_multi_objective.py

```diff
@@ -91,26 +91,31 @@ def _fast_non_dominated_sort(
                 len(penalty), len(objective_values)
             )
         )
-    nondomination_rank = np.zeros(len(objective_values), dtype=int)
+    nondomination_rank = np.full(len(objective_values), -1)
     is_nan = np.isnan(penalty)
+    n_below = n_below or len(objective_values)
 
     # First, we calculate the domination rank for feasible trials.
     is_feasible = np.logical_and(~is_nan, penalty <= 0)
-    ranks, feasible_bottom_rank = _calculate_nondomination_rank(
+    ranks, bottom_rank = _calculate_nondomination_rank(
         objective_values[is_feasible], n_below=n_below
     )
-    nondomination_rank[is_feasible] += ranks
+    nondomination_rank[is_feasible] += 1 + ranks
+    n_below -= np.count_nonzero(is_feasible)
 
     # Second, we calculate the domination rank for infeasible trials.
     is_infeasible = np.logical_and(~is_nan, penalty > 0)
-    ranks, infeasible_bottom_rank = _calculate_nondomination_rank(
-        penalty[is_infeasible, np.newaxis], n_below=n_below
+    ranks, bottom_rank = _calculate_nondomination_rank(
+        penalty[is_infeasible, np.newaxis], n_below=n_below, base_rank=bottom_rank + 1
     )
```

Comment:
For 1d array, `np.unique` is better.

```suggestion
    _, ranks = np.unique(penalty[is_infeasible], return_inverse=True)
    ranks += bottom_rank + 2
    bottom_rank = np.max(ranks)
```

Reply:
I replaced `_calculate_nondomination_rank` with `np.unique`.